ambry-network/src/main/java/com.github.ambry.network/BlockingChannelConnectionPool.java

```diff
@@ -223,6 +223,15 @@ public void destroyBlockingChannel(BlockingChannel blockingChannel) {
       synchronized (lock) {
         // decrement the number of connections to the host and port. we were not able to maintain the count
         numberOfConnections.decrementAndGet();
+        // at this point we are good to clean up the available connections since re-creation failed
+        do {
```

Comment:
Ah, actually the data structure here is not concurrent, so you will need the lock. My previous comment to change this to a poll and remove was also based on the assumption that this is a concurrent data structure.

You will need the lock here for sure to avoid concurrent access to this data structure. It will not solve the problem of connections getting added after they are removed (for the case mentioned above), but again that might be okay.